core/src/main/java/org/opensearch/sql/calcite/CalcitePlanContext.java

```diff
@@ -7,22 +7,52 @@
 
 import java.util.function.BiFunction;
 import lombok.Getter;
+import org.apache.calcite.jdbc.CalciteConnection;
+import org.apache.calcite.plan.Context;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptSchema;
+import org.apache.calcite.prepare.CalcitePrepareImpl;
 import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.server.CalciteServerStatement;
 import org.apache.calcite.tools.FrameworkConfig;
 import org.apache.calcite.tools.RelBuilder;
+import org.checkerframework.checker.nullness.qual.Nullable;
 import org.opensearch.sql.ast.expression.UnresolvedExpression;
 
 public class CalcitePlanContext {
 
+  public static class OSRelBuilder extends RelBuilder {
+
+    protected OSRelBuilder(
+        @Nullable Context context, RelOptCluster cluster, @Nullable RelOptSchema relOptSchema) {
+      super(context, cluster, relOptSchema);
+    }
+  }
+
   public FrameworkConfig config;
+  public CalciteConnection connection;
+  public CalciteServerStatement statement;
+  public final CalcitePrepareImpl prepare;
   public final RelBuilder relBuilder;
   public final ExtendedRexBuilder rexBuilder;
 
   @Getter private boolean isResolvingJoinCondition = false;
 
-  public CalcitePlanContext(FrameworkConfig config) {
+  public CalcitePlanContext(FrameworkConfig config, CalciteConnection connection) {
     this.config = config;
-    this.relBuilder = RelBuilder.create(config);
+    this.connection = connection;
+    try {
+      this.statement = connection.createStatement().unwrap(CalciteServerStatement.class);
+    } catch (Exception e) {
```

Comment:
minor: this try-catch is for unwrap or createStatement failure, better to separate the NPE out, or else mass of UT will go through via NPE here.
```
if (connection != null) {
  // try-catch block
}